### PR TITLE
Wider vector for FP16 RELU Grad on GPUs

### PR DESCRIPTION
**Pull request by kaixih**
This PR uses wider vector (8 FP16 values) for loading and storing in FP16 ReluGrad kernel to improve performance on Nvidia Ampere GPUs.

For older GPUs, the performance is expected to be unchanged.

fyi @nluehr

### CODE REVIEW

**Comment by gbaned** [comment]
@chsigg Can you please review this PR ? Thanks!

File: tensorflow/core/kernels/relu_op_gpu.cu.cc [hunk 1]
```diff
@@ -35,6 +35,7 @@ namespace tensorflow {
 
 typedef Eigen::GpuDevice GPUDevice;
 
+static constexpr int VectorSize = 8;
```
[the remark this answers is not in the document]
Comment (kaixih) [reply tensorflow/core/kernels/relu_op_gpu.cu.cc]: It is elements. Fixed.

File: tensorflow/core/kernels/relu_op_gpu.cu.cc [hunk 2]
```diff
@@ -93,6 +94,64 @@ __global__ void ReluGradHalfKernel(const Eigen::half* __restrict__ gradient,
...
+  int32 half8_count = count / VectorSize;
+  int32 index = blockIdx.x * blockDim.x + threadIdx.x;
+
+  if (index < half8_count) {
+    float4 gradient_h8 = reinterpret_cast<const float4*>(gradient)[index];
+    float4 feature_h8 = reinterpret_cast<const float4*>(feature)[index];
+    float4* p_backprop_h8 = reinterpret_cast<float4*>(backprop) + index;
+
+    half2 *gradient_h2 = reinterpret_cast<half2*>(&gradient_h8);
+    half2 *feature_h2 = reinterpret_cast<half2*>(&feature_h8);
```
[the remark this answers is not in the document]
Comment (kaixih) [reply tensorflow/core/kernels/relu_op_gpu.cu.cc:111]: I think we need to first cast `half` to `float4` for vector load and store. Then cast it to `half2` for actual computation.

[the remark this answers is not in the document]
Comment (kaixih) [reply tensorflow/core/kernels/relu_op_gpu.cu.cc:111]: I think this should be legit and valid. We did the same thing in the eigen library, like https://gitlab.com/libeigen/eigen/-/blob/master/Eigen/src/Core/arch/GPU/PacketMath.h#L533

**Review by sanjoy — APPROVED** [review]

File: tensorflow/core/kernels/relu_op_gpu.cu.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (timshen91) [reply tensorflow/core/kernels/relu_op_gpu.cu.cc:111]: But the Eigen code doesn't use float4. Does float4 require 128-bit alignment? Can you add an `assert()` check for the pointer alignment?

[the remark this answers is not in the document]
Comment (kaixih) [reply tensorflow/core/kernels/relu_op_gpu.cu.cc:111]: Yes, it requires 128-bit alignment. And we already have the checking code in the PR.

Eigen uses ulonglong2 as Packet4h2 which has the same length of float4.

**Comment by sanjoy** [comment]
This had to be rolled back because it broke the ROCm build.  [This](https://gist.github.com/sanjoy/62cb3862dc48c0666ae4ca712693667f) is the error message.  Please fix the compile error and I'll re-approve the PR.

Let me know if you need more help in narrowing down the issue.

**Comment by kaixih** [comment]
@sanjoy, I should be a simple fix. I pushed a new commit in my kaixih:relu_grad_vect_pr (https://github.com/kaixih/tensorflow/tree/relu_grad_vect_pr) branch but it doesn't show up here. Should I make another PR or what is your advice?

**Comment by sanjoy** [comment]
Creating a new PR should be simpler.

**Pull request merged**